src/NHibernate/Hql/Ast/ANTLR/QueryTranslatorImpl.cs

```diff
@@ -192,7 +198,7 @@ private void ErrorIfSelect()
 			}
 		}
 
-		public NHibernate.Loader.Loader Loader
+		public ILoader Loader
```

Comment:
Another breaking change. If willing to merge this for 5.5, the property `Loader` cannot change its type. It seems this change will need a new major version.

Reply:
What about this solution?

```suggestion
		//Interface:
		
		// 6.0 TODO : remove
		Loader.Loader Loader { get; }
			
		// 6.0 TODO : rename to Loader when overloaded property gets removed
		ILoader QueryLoader { get; }
		
		// Implementation:
		
		// 6.0 TODO : remove
		public Loader.Loader Loader => QueryLoader as Loader.Loader ?? throw new NotSupportedException("Loader is not of supported type.");

		// 6.0 TODO : rename to Loader when overloaded property gets removed
		public ILoader QueryLoader => _queryLoader;
```

I don't really like the cast, but don't see a way around it without breaking the interface.

Reply:
Yes, adding a member to an interface is a binary breaking change, due to the lack of a default implementation for external implementations of the interface.

But we are speaking of which interface? `ILoader`? Since it is introduced here, there is no trouble changing it until a release with it is done. And for 6.0, binary breaking changes will be allowed, so the todo would not be an issue.

So the new member `QueryLoader` would have to be added to a new "transitional interface".

About the cast, provided the code make sure to use the new `QueryLoader` property when handling an `ILoader`, instead of calling the old `Loader` property, it should be fine. For helping not missing this, the old property should be flagged with the `Obsolete` attribute, instructing to use `QueryLoader`. For avoiding a source breaking change with 6.0 event for those not calling any obsolete member, the TODO should be change to just "Remove Loader property" without the renaming todo on `QueryLoader`.

Reply:
No, the problem is the `public partial interface IQueryTranslator`. `IQueryTranslator` currently (Master) provides the property `Loader.Loader Loader { get; }`, which must be still supported if no breaking changes should be introduced.
So `QueryTranslatorImpl` has to implement this property, but the backing field is of type `IQueryLoader` now, due to the `IQueryLoaderFactory` introduction. This makes a cast from `IQueryLoader to Loader.Loader` necessary now.

**Honestly, I don't want to wait for a major release :)** So I would suggest the following change for `IQueryTranslator/QueryTranslatorImpl`:
```suggestion
		//IQueryTranslator:

		// 6.0 TODO : change type to ILoader
		Loader.Loader Loader { get; }
		
		// QueryTranslatorImpl:
		
		// 6.0 TODO : change type to ILoader
		public Loader.Loader Loader => QueryLoader as Loader.Loader ?? throw new NotSupportedException("Loader is not of supported type.");
```

**But the cast looks kinda like a code smell.** 

Alternativly, as u suggest we could add a new transitional interface which provides the above QueryLoader property of type ILoader, but some classes/interfaces are depending on `Loader.Loader` e.g.
![image](https://user-images.githubusercontent.com/47595100/207161290-ce72db78-5e34-4b81-8865-deb0818d3c59.png)
So a new transitional interface wouldn't help much.